File: 01122. Relative Sort Array/solution.js
/**
 * Problem: 1122. Relative Sort Array
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Sorts arr1 by the order of arr2; extras in ascending order
 *
 * @param {number[]} arr1 - Array to sort
 * @param {number[]} arr2 - Reference order
 *
 * @returns {number[]} - Sorted array
 */
const relativeSortArray = (arr1, arr2) => {
  // Create a mapping object to store the index of each element in arr2
  const orderMap = {}

  // Populate the mapping object with arr2 values and their indices (offset by 1 to avoid 0 as falsy)
  for (let i = 0; i < arr2.length; i++) orderMap[arr2[i]] = i + 1

  // Sort arr1 based on the custom order defined in arr2
  return arr1.sort((a, b) => {
    // If both elements exist in arr2, sort by their mapped order
    if (orderMap[a] && orderMap[b]) return orderMap[a] - orderMap[b]
    // If neither element exists in arr2, sort numerically in ascending order
    if (!orderMap[a] && !orderMap[b]) return a - b
    // If only 'a' does not exist in arr2, place it after 'b'
    if (!orderMap[a]) return 1

    // If only 'b' does not exist in arr2, place 'a' before 'b'
    return -1
  })
}

File: 01122. Relative Sort Array/solution.go
/**
 * Problem: 1122. Relative Sort Array
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func relativeSortArray(arr1 []int, arr2 []int) []int {
	// Initialize the result slice to store the final sorted array
	result := make([]int, 0)
	// Create a map to count occurrences of elements in arr2
	countMap := make(map[int]int)
	// Slice to store elements not present in arr2
	remains := make([]int, 0)

	// Initialize countMap with keys from arr2 and set their counts to 0
	for _, num := range arr2 {
		countMap[num] = 0
	}

	// Count occurrences of each number in arr1
	for _, num := range arr1 {
		count, exists := countMap[num]
		if exists {
			// If the number is in arr2, increment its count
			countMap[num] = count + 1
		} else {
			// If not, add it to remains
			remains = append(remains, num)
		}
	}

	// Sort the remains slice in ascending order
	slices.Sort(remains)

	// Append numbers from arr2 to result in the order of arr2, repeated by their counts
	for _, num := range arr2 {
		for i := 0; i < countMap[num]; i++ {
			result = append(result, num)
		}
	}

	// Append the sorted remains to the result
	result = append(result, remains...)

	// Return the final sorted array
	return result
}
